crates/leon/src/error.rs

```diff
@@ -0,0 +1,34 @@
+use miette::{Diagnostic, SourceSpan};
+use thiserror::Error;
+
+#[derive(Debug, Diagnostic, Error)]
+pub enum LeonError {
+    /// The template failed to parse.
+    #[error(transparent)]
+    InvalidTemplate(
+        #[diagnostic_source]
+        #[from]
+        ParseError,
+    ),
+
+    /// A key was missing from the provided values.
+    #[error("missing key `{0}`")]
+    MissingKey(String),
```

Comment:
Nit: If we use `Box<str>` here plus boxing `ParseError`, then size of `LeonError` will be the same as `String`.

Otherwise, even with `ParseError` boxed, its size will be `String` + `usize`.